<a href="https://colab.research.google.com/github/PhinyadaPhen/data_viz_2022/blob/main/Pandas101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd # import ฟังก์ชัน pandas 

.csv = Comma Separated Values

data.go.th

In [2]:
from google.colab import drive  # เชื่อม drive ของเรา
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os # จัดการเกี่ยวกับไฟล์ต่างๆ ในโฟล์เดอร์

In [4]:
path = '/content/drive/My Drive/dataviz_2021_data'  # ชี้ว่า data เราอยู่โฟล์เดอร์ไหน

In [5]:
# เอาที่อยู่ของไฟล์เชื่อมกับที่อยู่ของไฟล์เรา โดยกำหนดชื่่อไฟล์ที่เราต้องการ
covid_file_path = os.path.join(path,'pm-20-jan-2021.csv') 
print(covid_file_path)

/content/drive/My Drive/dataviz_2021_data/pm-20-jan-2021.csv


In [6]:
a = path + '/pm-18-jan-2021.csv' 
print(a)

/content/drive/My Drive/dataviz_2021_data/pm-18-jan-2021.csv


## load data to memory

In [7]:
data_covid = pd.read_csv(covid_file_path) # อ่าน .csv โดยใช้ pandas
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [8]:
data_covid.head() # .head() ใช้ดูข้อมูลในหัวตาราง

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

NaN = Not a Number

In [9]:
data_covid.head(20)

No.  ...                                               risk
0     1  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
1     2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2     3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
3     4  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
4     5  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
5     6  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
6     7  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7     8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
8     9  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
9    10  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
10   11  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11   12  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12   13  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
13   14  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
14   15  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
15   16  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
16   17  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
17   18  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
18   19  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
19   20  ...                      คนไทยเดินทางกลับจากต่างประเทศ

[20 rows x 10 columns]

In [10]:
data_covid.tail() # .tail() เรียกดูข้อมูลท้ายตาราง

No.  ...                                         risk
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5 rows x 10 columns]

In [11]:
data_covid.shape # แสดงขนาดข้อมูล

(12653, 10)

## ชี้ค่าในตาราง

#### basic

In [12]:
data_covid['province_of_onset'] # ชี้ค่าในตาราง = ชี้ column ธรรมดา

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [13]:
data_covid['risk'][12649] # ชี้ค่าในตารางก่อน ค่อยชี้ตัวที่ต้องการ  = column ตามด้วย row

'สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้'

#### .iloc

In [14]:
data_covid.iloc[4,7] # .iloc เป็นการเปลี่ยนการมองเห็นธรรมให้เป็น array

'กรุงเทพมหานคร'

# เรียนถึงนี่ 12 Jan 2022

## Table slicing

#### ตัดมาเฉพาะ column ที่ต้องการ

In [15]:
smaller_table = data_covid[['announce_date','province_of_onset','risk']]
smaller_table

announce_date  ...                                         risk
0         1/12/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1         1/17/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3         1/22/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4         1/24/2020  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...             ...  ...                                          ...
12648     1/20/2021  ...                             State Quarantine
12649     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651     1/20/2021  ...                                Cluster ระยอง
12652     1/20/2021  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 3 columns]

#### ตัดเฉพาะ rows ที่ต้องการ

##### แบบง่าย แต่ใช้การอะไรไม่ค่อยจะได้ (มองเป็น array แล้วตัดแบบ array)

In [16]:
data_covid.iloc[1:5,:]

No. announce_date  ... district_of_onset                              risk
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 10 columns]

แบบ advanced แต่ใช้ประโยชน์ได้เยอะกว่า (ตัดแบบใช้ logic query)

$pddf$[list_of_logical_values] คือการเลือก row ที่ logical values เป็น True

In [17]:
data_covid['province_of_onset'] # แสดง Column ของ province_of_onset ทั้งหมด

0        กรุงเทพมหานคร
1        กรุงเทพมหานคร
2               นครปฐม
3        กรุงเทพมหานคร
4        กรุงเทพมหานคร
             ...      
12648           ชลบุรี
12649            ระยอง
12650            ระยอง
12651            ระยอง
12652              ตาก
Name: province_of_onset, Length: 12653, dtype: object

In [18]:
# list ของค่า logic ต่างๆ โดยค่าเจอคำว่า "ขอนแก่น" จะ show "True"
data_covid['province_of_onset']=='ขอนแก่น'

0        False
1        False
2        False
3        False
4        False
         ...  
12648    False
12649    False
12650    False
12651    False
12652    False
Name: province_of_onset, Length: 12653, dtype: bool

In [19]:
data_covid[data_covid['province_of_onset']=='ขอนแก่น'] 

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

In [20]:
data_covid[data_covid['age']>=70] 

No.  ...                                               risk
1          2  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
7          8  ...                   คนต่างชาติเดินทางมาจากต่างประเทศ
24        25  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...
82        83  ...                                            สนามมวย
...      ...  ...                                                ...
11694  11695  ...      ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
11700  11701  ...        สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11730  11731  ...                                  Cluster สมุทรสาคร
11741  11742  ...                               อยู่ระหว่างการสอบสวน
12102  12103  ...                                  Cluster สมุทรสาคร

[245 rows x 10 columns]

การชี้ค่าในตารางด้วย data_covid[***x***] ค่า ***x*** สามารถมีค่าที่ใส่เข้าไปได้ 2 แบบ


1.   ชื่อ column
2.   ใส่ rows ที่ต้องการ (ใส่ list ที่มี***ขนาดเท่ากับจำนวนแถว*** แล้วบอกว่าเอา แถวไหนบ้างโดยใช้ logical expression (```True```/```False```))



In [21]:
eight_rows_covid = data_covid.iloc[:8,3:] # ตารางตัวอย่าง
eight_rows_covid

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
7  หญิง  73.0  ...            หัวหิน  คนต่างชาติเดินทางมาจากต่างประเทศ

[8 rows x 7 columns]

In [22]:
eight_rows_covid[[True,False,True,True,False,True,False,False]]   # 2.ใส่ rows ที่ต้องการ

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2  หญิง  73.0  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[4 rows x 7 columns]

In [23]:
eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร'  # สร้าง list ของ logical expression

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: province_of_onset, dtype: bool

In [24]:
eight_rows_covid[eight_rows_covid['province_of_onset'] == 'กรุงเทพมหานคร']  # เราเอา 2 บรรทัดบนมาบวกกัน

sex   age  ... district_of_onset                              risk
0  หญิง  61.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1  หญิง  74.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
3   ชาย  68.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4  หญิง  66.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
5  หญิง  33.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
6  หญิง  57.0  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[6 rows x 7 columns]

In [25]:
eight_rows_covid['district_of_onset'] == 'เมือง' # ค่าที่เราสนใจจะเป็น True เสมอ

0    False
1    False
2     True
3    False
4    False
5    False
6    False
7    False
Name: district_of_onset, dtype: bool

In [26]:
import numpy as np

$.isnull()$ ใช้ตรวจสอบว่าเป็น NaN หรือไม่

In [27]:
eight_rows_covid['district_of_onset'].isnull() # isnull() ใช้เช็คค่าว่าง

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7    False
Name: district_of_onset, dtype: bool

Quiz (ให้ใช้เฉพาะที่อาจารย์สอนไปแล้วในวิชานี้)

1.   คำนวณ อายุเฉลี่ย ของผู้ชาย ของข้อมูลทั้งหมด
2.   คำนวณ อายุเฉลี่ย ของผู้ชาย ของผู้ป่วยในจังหวัดขอนแก่น
3. หาจำนวนผู้ป่วยที่เป็นคนต่างชาติ ที่ไม่ใช่คนไืทย

In [28]:
dataQ = data_covid[['sex','age','province_of_onset','nationality']] # เลือก column ข้อมูลมาเฉพาะที่ต้องการใช้ 
dataQ # แสดงตาราง

sex   age province_of_onset nationality
0      หญิง  61.0     กรุงเทพมหานคร       China
1      หญิง  74.0     กรุงเทพมหานคร       China
2      หญิง  73.0            นครปฐม    Thailand
3       ชาย  68.0     กรุงเทพมหานคร       China
4      หญิง  66.0     กรุงเทพมหานคร       China
...     ...   ...               ...         ...
12648  หญิง  44.0            ชลบุรี    Thailand
12649  หญิง  52.0             ระยอง    Thailand
12650  หญิง  23.0             ระยอง    Thailand
12651  หญิง  29.0             ระยอง    Thailand
12652  หญิง  22.0               ตาก       Burma

[12653 rows x 4 columns]

In [29]:
# 1.คำนวณ อายุเฉลี่ย ของผู้ชาย ของข้อมูลทั้งหมด
male = dataQ[dataQ['sex']=='ชาย'] # เลือกมาเฉพาะค่าที่เป็นเพศชาย
male['age'] # หาอายุของเพศชาย

3        68.0
8        63.0
10       33.0
11       61.0
12        6.0
         ... 
12640    66.0
12641    29.0
12643    39.0
12646    18.0
12647    42.0
Name: age, Length: 4819, dtype: float64

In [30]:
# 1.คำนวณ อายุเฉลี่ย ของผู้ชาย ของข้อมูลทั้งหมด

sum = 0 # sum มารับ โดยให้เป็น 0 เนื่องจากจะนำมารับค่าทั้งหมดที่จะบวกกัน
N = 0 # N นับจำนวนคนทั้งหมด เริ่มที่ 0
for a in male['age']: # นำค่าอายุของเพศชายเข้ามาเพื่อวนดูค่าอายุ 
    
    if a > 0: # ถ้า a > 0 แสดงว่าเป็นอายุของคน ดังนั้นจะนำมานับ
        sum += a # sum = sum + a # คำนวณหาอายุของเพศชาย ซึ่งถ้า a เป็นอายุคนก็จะถูกนำมาบวกกันเรื่อยๆ
        N += 1 # คำนวณหาจำนวนผู้ชายทั้งหมดโดยเพิ่มทีละ 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยชาย {sum/N}') # แสดงค่าเฉลี่ยของอายุผู้ป่วยชาย โดยให้ sum = อายุรวมทั้งหมดของผู้ป่วยชาย หารด้วย N = จำนวนผู้ป่วยชายทั้งหมด

อายุเฉลี่ย ของ ผู้ป่วยชาย 39.46076940331987


In [31]:
# 2.คำนวณ อายุเฉลี่ย ของผู้ชาย ของผู้ป่วยในจังหวัดขอนแก่น

male_KK = male[male['province_of_onset']=='ขอนแก่น'] # เลือกค่าที่เป็นเพศชาย ที่อยู่ในจังหวัดขอนแก่น

sum = 0 # sum มารับ โดยให้เป็น 0 เนื่องจากจะนำมารับค่าทั้งหมดที่จะบวกกัน
N = 0 # N นับจำนวนคนทั้งหมด เริ่มที่ 0
for a in male_KK['age']: # นำค่าอายุของเพศชายเข้ามาเพื่อวนดูค่าอายุ 
    
    if a > 0: # ถ้า a > 0 แสดงว่าเป็นอายุของคน ดังนั้นจะนำมานับ
        sum += a # sum = sum + a # คำนวณหาอายุของเพศชาย ซึ่งถ้า a เป็นอายุคนก็จะถูกนำมาบวกกันเรื่อยๆ
        N += 1 # คำนวณหาจำนวนผู้ชายทั้งหมดโดยเพิ่มทีละ 1

print(f'อายุเฉลี่ย ของ ผู้ป่วยชาย ในจังหวัดขอนแก่น {sum/N}') # แสดงค่าเฉลี่ยของอายุผู้ป่วยชายในจังหวัดขอนแก่น โดยให้ sum = อายุรวมของผู้ป่วยชายในจังหวัดขอนแก่น หารด้วย N = จำนวนผู้ป่วยชายทั้งหมด

อายุเฉลี่ย ของ ผู้ป่วยชาย ในจังหวัดขอนแก่น 26.571428571428573


In [32]:
dataN = data_covid[['nationality']] # เลือกข้อมูลที่ต้องการนำมาใช้งาน โดยเลือกเฉพาะ column 'nationality' ในการนำมาหาผู้ป่วยที่ไม่ใช่คนไทย
dataN # แสดงค่าข้อมูลว่ามีประเทศอะไรบ้าง

nationality
0           China
1           China
2        Thailand
3           China
4           China
...           ...
12648    Thailand
12649    Thailand
12650    Thailand
12651    Thailand
12652       Burma

[12653 rows x 1 columns]

In [33]:
# 3.หาจำนวนผู้ป่วยที่เป็นคน "ไม่ใช่คนไทย"

patient = dataN[dataN['nationality'] != 'Thailand'] # หาผู้ป่วยที่ไม่ใช่คนไทย จาก column 'nationality' เลือกโดย ไม่เอาไทย 
patient # แสดงข้อมูลผู้ป่วยที่ไม่ใช่คนไทย

nationality
0           China
1           China
3           China
4           China
5           China
...           ...
12614       Burma
12615       Burma
12617     Ireland
12618    Pakistan
12652       Burma

[5340 rows x 1 columns]

In [34]:
# 3.หาจำนวนผู้ป่วยที่เป็นคน "ไม่ใช่คนไทย"

patient = dataN[dataN['nationality'] != 'Thailand'] # หาผู้ป่วยที่ไม่ใช่คนไทย จาก column 'nationality' เลือกโดย ไม่เอาไทย
total = patient.count() # นำค่าที่ไม่มีคนไทยมาหาจำนวนผู้ป่วยทั้งหมด โดยใช้คำสั่ง count ในการนับจำนวน #count จะนับเป็นจำนวนครั้งเมื่อค้นเจอค่าข้อมูล
total # แสดงค่าออกมา

nationality    4129
dtype: int64

จากการคำนวณหาจำนวนผู้ป่วยที่ไม่ใช่คนไทย โดยใช้คำสั่ง count จะเห็นได้ว่า ผู้ป่วยที่ไม่ใช่คนไทยมีจำนวน 4129 คน (ดูจาก column 'nationality')

เฉลยวิธีหาข้อ(3.) แบบอาจารย์ 

In [35]:
import numpy as np

In [36]:
np.sum(dataQ['nationality'].dropna() != 'Thailand')

4129

In [37]:
this_data = data_covid[['sex','age','province_of_onset']]

In [38]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

การจัดการ Missing Value

- ลบ record ที่เป็น missing 
- แทนที่ ค่า missing ด้วยค่าที่เหมาะสม mean, default, category-unknown
- ใช้ ค่าจาก columns อื่นๆ ช่วยประมาณค่า ค่าใน column ที่หายไป (regression, deep learning, etc.)

### ลบ record (dropna)

missing = ```None``` ```NA```(not autorized) ```NaN``` (not a number)

In [39]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [40]:
this_data.shape #(12423, 3)

(12653, 3)

In [41]:
this_data.dropna()

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8478 rows x 3 columns]

In [42]:
print(data_covid.shape)
print(data_covid.dropna().shape)

(12653, 10)
(6655, 10)


In [43]:
this_data.dropna(subset=['age','sex'])

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

In [44]:
print(this_data.dropna().shape)
print(this_data.dropna(subset=['sex','age']).shape)

(8478, 3)
(8880, 3)


# ตรวจสอบ Missing ```.isnull()```

In [45]:
this_data.isnull()

sex    age  province_of_onset
0      False  False              False
1      False  False              False
2      False  False              False
3      False  False              False
4      False  False              False
...      ...    ...                ...
12648  False  False              False
12649  False  False              False
12650  False  False              False
12651  False  False              False
12652  False  False              False

[12653 rows x 3 columns]

In [46]:
this_data.isnull().any() # .any() เอาค่าความจริงใน column มา OR กัน

sex                  True
age                  True
province_of_onset    True
dtype: bool

In [47]:
this_data.isnull().all() # .all() เอาค่าความจริงใน column มา AND กัน

sex                  False
age                  False
province_of_onset    False
dtype: bool

In [48]:
this_data.isnull().any().any()

True

In [49]:
this_data.dropna(subset=['sex','age']).isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [50]:
this_data.dropna(subset=['sex','age'])

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[8880 rows x 3 columns]

### 1. เพื่อให้ได้ผลลัพธ์การจัดการ ```dropna()``` ต้องเอาตัวแปรมารับ

In [51]:
this_data_nonull = this_data.dropna(subset=['sex','age'])

In [52]:
this_data_nonull.isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

### 2. Update ตารางด้วยตัวแปร  ```inplace```

In [53]:
this_data.dropna(subset=['sex','age'],inplace=True) 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [54]:
this_data.isnull().any()

sex                  False
age                  False
province_of_onset     True
dtype: bool

In [55]:
aa = this_data.dropna(subset=['sex','age'],inplace=True) 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [56]:
aa

### แทน missing ด้วยค่าที่เหมาะสม (```fillna()```)

In [57]:
this_data = data_covid[['sex','age','province_of_onset']]

In [58]:
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [59]:
print(f'จำนวน record ก่อน drop missing ใน province {this_data.shape[0]}')
print(f'จำนวน record หลัง drop missing ใน province {this_data.dropna(subset=["province_of_onset"]).shape[0]}')

จำนวน record ก่อน drop missing ใน province 12653
จำนวน record หลัง drop missing ใน province 10662


### 1. แทนด้วย class ใหม่

In [60]:
this_data_updated = this_data.fillna(value={'sex':'ไม่รู้', 'age':-99, 'province_of_onset':'ไม่รู้'})
this_data_updated.head()

sex   age province_of_onset
0  หญิง  61.0     กรุงเทพมหานคร
1  หญิง  74.0     กรุงเทพมหานคร
2  หญิง  73.0            นครปฐม
3   ชาย  68.0     กรุงเทพมหานคร
4  หญิง  66.0     กรุงเทพมหานคร

In [61]:
this_data_updated.tail(15)

sex   age province_of_onset
12638  ไม่รู้ -99.0         สมุทรสาคร
12639  ไม่รู้ -99.0         สมุทรสาคร
12640     ชาย  66.0   พระนครศรีอยุธยา
12641     ชาย  29.0          ปทุมธานี
12642    หญิง  50.0           ราชบุรี
12643     ชาย  39.0              สตูล
12644    หญิง  49.0       สมุทรปราการ
12645    หญิง  44.0            ชลบุรี
12646     ชาย  18.0       สมุทรปราการ
12647     ชาย  42.0     กรุงเทพมหานคร
12648    หญิง  44.0            ชลบุรี
12649    หญิง  52.0             ระยอง
12650    หญิง  23.0             ระยอง
12651    หญิง  29.0             ระยอง
12652    หญิง  22.0               ตาก

In [62]:
this_data_updated.isnull().any()

sex                  False
age                  False
province_of_onset    False
dtype: bool

### 2. แทนด้วยค่าที่เหมาะสม

*   ตามใจคนทำ
*   ค่ากลาง

#### ตามใจคนทำ

In [63]:
data_covid.head()

No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [64]:
data_covid_fillasiwant = data_covid.fillna(value={'sex':'หญิง'})  #เรามีข้อมูลว่า เพศหญิงมีเยอะกว่าเพศชาย

In [65]:
data_covid_fillasiwant.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                      False
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

#### แทนด้วยค่ากลาง

In [66]:
data_covid['age']

0        61.0
1        74.0
2        73.0
3        68.0
4        66.0
         ... 
12648    44.0
12649    52.0
12650    23.0
12651    29.0
12652    22.0
Name: age, Length: 12653, dtype: float64

In [67]:
data_covid['age'].mean()

38.054748339151

In [68]:
data_covid_fillwithmean = data_covid.fillna(value={'age':data_covid['age'].mean()})

In [69]:
data_covid_fillwithmean.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                       True
age                      False
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

### การวนลูป record ในตาราง  (```.iterrows```)

In [70]:
this_data = data_covid[['sex','age','province_of_onset']]
this_data

sex   age province_of_onset
0      หญิง  61.0     กรุงเทพมหานคร
1      หญิง  74.0     กรุงเทพมหานคร
2      หญิง  73.0            นครปฐม
3       ชาย  68.0     กรุงเทพมหานคร
4      หญิง  66.0     กรุงเทพมหานคร
...     ...   ...               ...
12648  หญิง  44.0            ชลบุรี
12649  หญิง  52.0             ระยอง
12650  หญิง  23.0             ระยอง
12651  หญิง  29.0             ระยอง
12652  หญิง  22.0               ตาก

[12653 rows x 3 columns]

In [71]:
for each_row in this_data.iterrows():
    # print(each_row)
    # print(each_row[1])
    # print(each_row[1]['sex'])
    if (each_row[1]['age'] <= 20) and (each_row[1]['province_of_onset'] == 'ขอนแก่น'):
        print(each_row)

(1466, sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object)
(9333, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object)
(9334, sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object)
(10610, sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object)
(11517, sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object)
(11697, sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object)


### การวนลูป แบบมองตารางแพนด้าส์(pandas dataframe) เป็น numpy array  (```.iloc```)

In [72]:
for each_row in range(this_data.shape[0]):
    if (this_data.iloc[each_row,1] <= 20) and (this_data.iloc[each_row,2] == 'ขอนแก่น'):  #each_row คือ ตัวเลข
        print(each_row)
        print(this_data.iloc[each_row,:])   

1466
sex                      ชาย
age                       19
province_of_onset    ขอนแก่น
Name: 1466, dtype: object
9333
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 9333, dtype: object
9334
sex                      ชาย
age                       20
province_of_onset    ขอนแก่น
Name: 9334, dtype: object
10610
sex                      ชาย
age                       17
province_of_onset    ขอนแก่น
Name: 10610, dtype: object
11517
sex                      ชาย
age                       12
province_of_onset    ขอนแก่น
Name: 11517, dtype: object
11697
sex                     หญิง
age                       17
province_of_onset    ขอนแก่น
Name: 11697, dtype: object


## Quiz ตัดตารางออกมาเฉพาะปี 2021 ```announce_date``` ในปี 2021 

**Hint**
*   วนลูปหา index ของปี 2021
*   ตัดตารางมาเฉพาะ ปี 2021



In [73]:
TF=list()
for each_row in data_covid.iterrows():
    if each_row[1]['announce_date'].split('/')[2] == '2021':
        TF.append(True)
    else:
        TF.append(False)

data_covid[TF].head()

No.  ...                                               risk
6884  6885  ...                                  Cluster สมุทรสาคร
6885  6886  ...                                      Cluster ระยอง
6886  6887  ...                                        สถานบันเทิง
6887  6888  ...                                  Cluster สมุทรสาคร
6888  6889  ...  อาชีพเสี่ยง เช่น ทำงานในสถานที่แออัด หรือทำงาน...

[5 rows x 10 columns]

In [78]:
TF

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,


### HW กลุ่ม 2 || เลือกข้อมูลมาเฉพาะวัน announce_date เป็นวันที่ 1-20  

In [74]:
ad =list() # สร้าง list เปล่าขึ้นมา แล้วนำไปเก็บไว้ที่ตัวแปร ad
for each_row in data_covid.iterrows(): # ไปดูทุกแถวในตาราง data_covid 
    if each_row[1]['announce_date'].split('/')[1] <= '20': #เป็นการดูแต่ละแถว, [1]['announce_date'] คือการเรียกดูข้อมูลเริ่มจากตัวที่ 1 ของแต่ละแถว และดูที่คอลัมน์ announce_date และแบ่งค่าข้อมูลโดยตัดตามเครื่องหมาย / เมื่อตัดแล้ว ตัวที่1 ของคอลัมน์ announce_date ต้องมีค่าน้อยกว่าหรือเท่ากับ 20
        ad.append(True) # ถ้าข้อมูลตรงตามเงื่อนไข  จะทำการเพิ่มข้อมูลเข้าไปใน list ad
    else:
        ad.append(False) # ถ้าข้อมูลไม่ตรงตามเงื่อนไข จะไม่เพิ่มข้อมูลเข้าไปใน list

data_covid[ad] # แสดงค่าข้อมูลตาราง data_covid 

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
32        33  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
33        34  ...              บุคลากรด้านการแพทย์และสาธารณสุข
34        35  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5135 rows x 10 columns]

# Function ตัวช่วยใน pandas

### ```.describe()``` คำนวณค่าทางสถิติของข้อมูลที่เป็นตัวเลข

In [75]:
data_covid.describe()

No.          age
count  12653.000000  8881.000000
mean    6327.000000    38.054748
std     3652.750813    15.125178
min        1.000000     0.100000
25%     3164.000000    27.000000
50%     6327.000000    36.000000
75%     9490.000000    48.000000
max    12653.000000    97.000000

### ```.mean()``` คำนวณค่าเฉลี่ยของข้อมูลโดยไม่สนใจ missing 

In [76]:
data_covid[data_covid['sex']=='ชาย']['age'].mean()

39.46076940331987

### ```.isnull()```

In [77]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
12648  False          False  ...              False  False
12649  False          False  ...              False  False
12650  False          False  ...              False  False
12651  False          False  ...              False  False
12652  False          False  ...              False  False

[12653 rows x 10 columns]

## next
* ต่อตารางแกน X แกน y
* .groupby
* save table

# next
*   .describe
*   .mean
*   HW วนลูป missing .isnull
*   ต่อตารางแกน X แกน y
*   .groupby
*   save table